### PR TITLE
Offer to install Posh-Git in the installer

### PR DESCRIPTION
**Pull request by dscho**
[Posh-Git](https://github.com/dahlbyk/posh-git) is a very neat PowerShell module that serves as a PowerShell equivalent to Git's Bash prompt and tab completion.

Let's offer it as an optional component; When selected, it will be installed from the PSGallery for all users and will then also be added to the profile.

This addresses https://github.com/git-for-windows/git/issues/1384

/cc @dahlbyk

### CODE REVIEW

File: installer/install.iss [hunk 1]
```diff
@@ -3366,6 +3368,39 @@ begin
...
+            'if ($policy -ne """Trusted""") {' +
+            ' Set-PSRepository -Name PSGallery -InstallationPolicy Trusted '+
+            '} ' +
+            'Install-Module -Repository PSGallery -Scope AllUsers posh-git; ' +
+            '$res=$?; ' +
+            'if ($policy -ne """Trusted""") {' +
+            ' Set-PSRepository -Name PSGallery -InstallationPolicy $policy ' +
+            '} ' +
+            'if ($res) {' +
+            ' Add-PoshGitToProfile -AllUsers; ' +
```
Comment (dahlbyk) [inline installer/install.iss:3393]: This includes a check if posh-git is already imported in _any_ of the current user's profiles, which would prevent the expected system-wide install here. It would seem reasonable for posh-git to be changed to only test the file to which the import would be added.

However, when testing this I also noticed that importing posh-git from two different locations (e.g. from the system install and then from a user install) can lead to unexpected behavior. So maybe better to err on the side of leaving all profiles alone if there's evidence of posh-git?

(No change required, here. Just thinking out loud about how posh-git should behave.)

**Review by dahlbyk — COMMENTED** [review]
Pulling this over from https://github.com/git-for-windows/git/issues/1384#issuecomment-987890826:

> Should we uninstall Posh-Git when uninstalling Git for Windows?

Importing posh-git without a `git` command available [logs a warning](https://github.com/dahlbyk/posh-git/blob/b79c2dc39c9387847642bc3b38fa2186b29f6113/src/CheckRequirements.ps1#L5-L9). Not fatal, but not ideal; better to be able to uninstall.

Can you tell if Git for Windows was installed with the `poshgit` component? Ideally we'd leave the module alone if it was installed some other way.

> If so, how do I remove it from the users' profile, programmatically?

Relevant code exists as of https://github.com/dahlbyk/posh-git/pull/358; we can clean that up and export `Remove-PoshGitFromProfile -AllUsers`.

**Comment by dahlbyk** [comment]
cc @rkeithhill @lzybkr

**Comment by dscho** [comment]
> Can you tell if Git for Windows was installed with the `poshgit` component?

Yes. We simply ask `IsComponentInstalled('poshgit')` (see the corresponding [uninstall step for the auto-updater](https://github.com/git-for-windows/build-extra/blob/bced1bdc74d13cec09d966c1e0f6aef103548ce0/installer/install.iss#L3758-L3759)).

> > If so, how do I remove it from the users' profile, programmatically?
> 
> Relevant code exists as of [dahlbyk/posh-git#358](https://github.com/dahlbyk/posh-git/pull/358); we can clean that up and export `Remove-PoshGitFromProfile -AllUsers`.

That would be nice! Do you think you can do it? I ask because 1) I am almost a PowerShell illiterate, and 2) I am currently overwhelmed with [working on an unwelcome blocker](https://github.com/git-for-windows/git/issues/3579#issuecomment-989286753).

File: installer/install.iss [hunk 2]
```diff
@@ -3366,6 +3368,39 @@ begin
...
+        Cmd:='"'+ExpandConstant('{sysnative}\WindowsPowerShell\v1.0\powershell.exe')+'"'+
+            ' -ExecutionPolicy Bypass -NoLogo -NonInteractive -WindowStyle Hidden -command "'+
+            'if (!(Get-PackageProvider -Name NuGet)) {'+
+            ' Install-PackageProvider -Name NuGet -Force ' +
+            '} ' +
+            '$policy = (Get-PSRepository -Name PSGallery).InstallationPolicy; ' +
+            'if ($policy -ne """Trusted""") {' +
+            ' Set-PSRepository -Name PSGallery -InstallationPolicy Trusted '+
+            '} ' +
+            'Install-Module -Repository PSGallery -Scope AllUsers posh-git; ' +
```
[the remark this answers is not in the document]
Comment (dscho) [reply installer/install.iss:3387]: Makes sense.

File: installer/install.iss [hunk 3]
```diff
@@ -3366,6 +3368,39 @@ begin
...
 
+    {
+        Install Posh-Git from the PSGallery
+    }
+
+    if (IsComponentSelected('poshgit')) then begin
+        WizardForm.StatusLabel.Caption:='Installing Posh-Git from the PSGallery';
+        // Must use the sysnative version of PowerShell, otherwise will target the wrong profile
+        Cmd:='"'+ExpandConstant('{sysnative}\WindowsPowerShell\v1.0\powershell.exe')+'"'+
+            ' -ExecutionPolicy Bypass -NoLogo -NonInteractive -WindowStyle Hidden -command "'+
```
[the remark this answers is not in the document]
Comment (dscho) [reply installer/install.iss:3379]: Sure!

**Comment by dscho** [comment]
@dahlbyk I updated the commit (anticipating suppport for `Remove-PoshGitFromProfile -AllUsers`). Could I trouble you to review https://github.com/git-for-windows/build-extra/compare/489804d4184902ef9949eacc12823bd7f3d565f2..ec4db6d37c257d4e61d1cf877d76c8f9888bd53e?

**Comment by dscho** [comment]
> > Relevant code exists as of [dahlbyk/posh-git#358](https://github.com/dahlbyk/posh-git/pull/358); we can clean that up and export `Remove-PoshGitFromProfile -AllUsers`.
> 
> That would be nice! Do you think you can do it?

Actually, let me give it a shot (while [waiting for the MSYS2 runtime to compile](https://xkcd.com/303/)).

File: installer/install.iss [hunk 4]
```diff
@@ -3812,4 +3850,24 @@ begin
...
+        Remove posh-git
+    }
+
+    if (IsComponentSelected('poshgit')) then begin
+        Cmd:='"'+ExpandConstant('{sysnative}\WindowsPowerShell\v1.0\powershell.exe')+'"'+
+            ' -ExecutionPolicy Bypass -NoProfile -NoLogo -NonInteractive -WindowStyle Hidden -command "'+
+            'Uninstall-Package posh-git -Scope AllUsers -ErrorAction SilentlyContinue; ' +
+            '$res=$?; ' +
+            'if ($res) {' +
+            ' Remove-PoshGitFromProfile -AllUsers; ' +
```
Comment (dahlbyk) [inline installer/install.iss:3864]: Since this function will be provided by the module, should we uninstall after it has been removed from profile?

Comment (dscho) [reply installer/install.iss:3864]: But of course :grin:

**Comment by dscho** [comment]
> > > Relevant code exists as of [dahlbyk/posh-git#358](https://github.com/dahlbyk/posh-git/pull/358); we can clean that up and export `Remove-PoshGitFromProfile -AllUsers`.
> > 
> > 
> > That would be nice! Do you think you can do it?
> 
> Actually, let me give it a shot (while [waiting for the MSYS2 runtime to compile](https://xkcd.com/303/)).

@dahlbyk here you are: https://github.com/dahlbyk/posh-git/pull/877

BTW I also wanted to verify that it passes the test suite, but it failed. The test suite failed also without my changes, though, so I opened https://github.com/dahlbyk/posh-git/pull/878 which I hope you find the time to look into?

File: installer/install.iss (hunk 2 above)
[the remark this answers is not in the document]
Comment (dscho) [reply installer/install.iss:3388]: Do we need to test for that, or do we simply just uninstall? I am of two minds there. If a user uninstalled Posh-Git manually, they would probably want to be notified that Git still thinks it should be installed, right?

**Comment by dscho** [comment]
> @dahlbyk here you are: [dahlbyk/posh-git#877](https://github.com/dahlbyk/posh-git/pull/877)

Oh, and of course that PR (as well as a Posh-Git release after merging it) is a blocker for this here PR.

**Comment by dscho** [comment]
Git for Windows v2.35.0-rc0 is due either tonight or tomorrow morning. I had hoped to get Posh-Git into that release, but it is cutting it too close now even if we can get the project to offer Posh-Git via Git for Windows moving forward again.

**Comment by dahlbyk** [comment]
> Git for Windows v2.35.0-rc0 is due either tonight or tomorrow morning. I had hoped to get Posh-Git into that release, but it is cutting it too close now even if we can get the project to offer Posh-Git via Git for Windows moving forward again.

Agree. Still ramping up after disconnected entirely over break, but hope to get back into this soon.

**Comment by dscho** [comment]
-rc2 is due Wednesday. We've time to get this integrated until then, or after 2.35.0.

**Comment by dahlbyk** [comment]
> -rc2 is due Wednesday. We've time to get this integrated until then, or after 2.35.0.

If you don't see progress on this tonight, let's assume we're waiting for vNext.

**Comment by dscho** [comment]
Even if it is too late for yet another Git for Windows version, let's push this forward so that we do not miss the next train, too. First stop: tell me how I can fix the CI runs (which _all_ fail right now), then take a sweep at the open PRs/issues.

**Comment by dscho** [comment]
Git for Windows v2.35.0 was released yesterday. @dahlbyk I would like to push for Posh-Git to have a working GitHub workflow, the to accept the two PRs I opened, and then to have a new release so that we can go forward with this here PR.

**Comment by dahlbyk** [comment]
> Git for Windows v2.35.0 was released yesterday. @dahlbyk I would like to push for Posh-Git to have a working GitHub workflow, the to accept the two PRs I opened, and then to have a new release so that we can go forward with this here PR.

💯 🔜

**Review by dahlbyk — COMMENTED** [review]
So now that you waited months for `Remove-PoshGitFromProfile`, I'm more and more convinced we should just leave all profiles alone and just find a way to communicate that the user can `Add-PoshGitToProfile` from any PowerShell host to auto-import the system-installed module.

- Profile scripts don't work without changing Execution Policy; we could have the installer offer to change that for folks (need to give options), but I wouldn't want to do it silently
- Installing to `CurrentHost` means any host except the default (e.g. in VS Code) won't have posh-git, so the user will need to know about `Add-PoshGitToProfile` to turn it on there
  - Or we could use `AllHosts`, but there may be hosts that don't work well with posh-git and you can't opt out
- PowerShell Core uses entirely different `$PROFILE`s, so we could either install there _too_ or we're back to needing education about `Add-PoshGitToProfile`

In other words, this could well turn into a whole new page of the installer devoted to configuring the PowerShell environment:
- Execution Policy?
- Current or All Users?
- Windows PowerShell: Default Host, All Hosts, or none?
- PowerShell Core (if `pwsh.exe`): Default Host, All Hosts, or none?

Or we could just figure out somewhere for the installer to link to Learn More. 🤷‍♂️

**Comment by dscho** [comment]
> I'm more and more convinced we should just leave all profiles alone and just find a way to communicate that the user can `Add-PoshGitToProfile` from any PowerShell host to auto-import the system-installed module.

I'm not sure that I buy this argument. The problem I see here is that a user installed Git for Windows, clicked the option to install `posh-git`, then uninstalls Git for Windows (believing that it would clean up `posh-git`, too), and is then presented with a broken PowerShell profile (that may or may not result in a totally messed-up prompt).

**Comment by dahlbyk** [comment]
> > I'm more and more convinced we should just leave all profiles alone and just find a way to communicate that the user can `Add-PoshGitToProfile` from any PowerShell host to auto-import the system-installed module.
> 
> I'm not sure that I buy this argument. The problem I see here is that a user installed Git for Windows, clicked the option to install `posh-git`, then uninstalls Git for Windows (believing that it would clean up `posh-git`, too), and is then presented with a broken PowerShell profile (that may or may not result in a totally messed-up prompt).

I'm not sure we can avoid this. Even setting aside PS Core, a different user could install the system posh-git from their user-specific profile and we can't reasonably fix that on their behalf.

The simplest we can make this:
- Add to to All Users profile and only remove from All Users, if present
- Current Host is probably safer than All Hosts
- Try to add/remove for `pwsh` profile in addition to `powershell`, if command exists

Conflicting installs and other profiles are on the user to manage.

How easy would it be to make adding to profile an option, either binary or pick All/Current User/Host? Would the uninstall remember what was chosen?

**Comment by dahlbyk** [comment]
> > How easy would it be to make adding to profile an option, either binary or pick All/Current User/Host?
> 
> That could be a checkbox. [Here](https://jrsoftware.org/ishelp/topic_componentssection.htm) and [here](https://jrsoftware.org/ishelp/topic_componentstasksparams.htm) are valuable pages about this. It would look somewhat like this:

The `exclusive` component flag could let us model both All/Current Host and All/Current User. 🤔 

- ☐ Install Posh-Git from PSGallery
  - ☐ Install for Current User `exclusive`
  - ☐ Install for All Users `exclusive dontinheritcheck`
  - ☐ Add to Profile? (see Add-PoshGitToProfile) `disablenouninstallwarning`
    - ☐ Default Host `exclusive`
    - ☐ All Hosts `exclusive dontinheritcheck`

**Comment by dscho** [comment]
Sounds good, as long as `posh-git` delivers the relevant mode of `Remove-PoshGitFromProfile`.

**Comment by dscho** [comment]
Oh well. The fact that this has been open for such a long time without anybody even _trying_ to put some entitled pressure on me shows that I need to let go of this.
